mongodb-chunk-imbalance.md (#102)
percona-platform/portal-doc: docs/advisors/checks/mongodb-chunk-imbalance.md

> # MongoDB Sharding - Chunk Imbalance Across Shards

Reviewer: ```suggestion
# MongoDB sharding - chunk imbalance across shards
```

## Description
This check warns if the chunks are imbalanced across shards.

## Resolution

In a sharded cluster, chunk imbalances can occur when the data is not evenly distributed among the shards. This can lead to some shards having more chunks than others. This can cause performance issues and slower query times.

There are a few possible reasons for chunk imbalance:

- **Poor shard key selection:** If the shard key is not chosen properly, some shards may end up with a larger portion of the data than others. For example, if the shard key is based on a timestamp and the data is inserted in a sequential manner, this can create hotspots due to busier application write periods. 

> - **Jumbo Chunks:** Another cause of chunk imbalances in a sharded cluster can be due to jumbo chunks. Jumbo chunks are chunks that have grown beyond the maximum size that is allowed by MongoDB. When a chunk becomes jumbo, MongoDB cannot split it further automatically and the balancer won’t distribute the associated data across the shards. Jumbo chunks are often caused by low cardinality or too high frequency of elements in a shard key.

Reviewer: ```suggestion
- **Jumbo chunks:** Jumbo chunks can be another cause of chunk imbalances in a sharded cluster . Jumbo chunks are chunks that have grown beyond the maximum size that is allowed by MongoDB. When a chunk becomes jumbo, MongoDB cannot split it further automatically and the balancer won’t distribute the associated data across the shards. Jumbo chunks are often caused by low cardinality or too high frequency of elements in a shard key.
```

To address chunk imbalance, you can:

- **Good shard key selection:** 

> The choice of shard key affects the creation and distribution of chunks across the available shards. While selecting the shard key, consider below points -

Reviewer: ```suggestion
The choice of shard key affects the creation and distribution of chunks across the available shards. While selecting the shard key, consider below points:

```

1. High Cardinality.
2. Non-monotonic is nature.
3. It should be used in most of your queries.
4. Reads should be done from particular shards.
5. Writes should get written across all shards.

Starting in MongoDB 5.0, you can reshard a collection by changing a collection's shard key. Follow the [MongoDB Documentation](https://www.mongodb.com/docs/manual/core/sharding-reshard-a-collection/#std-label-sharding-resharding) on how to change the shard key.

> - **Clearing the Jumbo Chunks:**

Reviewer: ```suggestion
- **Clear the jumbo chunks:**
```

# MongoDB sharding - chunk imbalance across shards

## Description
This check warns if the chunks are imbalanced across shards.

> ## Resolution

Reviewer: ```suggestion

```

> In a sharded cluster, chunk imbalances can occur when the data is not evenly distributed among the shards. This can lead to some shards having more chunks than others. This can cause performance issues and slower query times.

Reviewer 2: ```suggestion
In a sharded cluster, chunk imbalances can occur when the data is not evenly distributed among the shards. This can lead to some shards having more chunks than others, which can cause performance issues and slower query times.
```

There are a few possible reasons for chunk imbalance:

- **Poor shard key selection:** If the shard key is not chosen properly, some shards may end up with a larger portion of the data than others. For example, if the shard key is based on a timestamp and the data is inserted in a sequential manner, this can create hotspots due to busier application write periods. 

- **Jumbo chunks:** Jumbo chunks can be another cause of chunk imbalances in a sharded cluster . Jumbo chunks are chunks that have grown beyond the maximum size that is allowed by MongoDB. When a chunk becomes jumbo, MongoDB cannot split it further automatically and the balancer won’t distribute the associated data across the shards. Jumbo chunks are often caused by low cardinality or too high frequency of elements in a shard key.

> 

Author: @nastena1606 Yes we can.

> To address chunk imbalance, you can:

Reviewer: ```suggestion
## Resolution

To address chunk imbalance, you can:
```

- **Select a good shard key:** 

> The choice of shard key affects the creation and distribution of chunks across the available shards. While selecting the shard key, consider below points:

Reviewer 2: ```suggestion
The choice of shard key impacts the way chunks are created and distributed across the available shards. While selecting the shard key, consider the following factors:
```

> 1. High Cardinality.

Reviewer 2: Numbered list shows chronology, imploying there's a procedure that should be followed in the order described. Would it be better to use bullet points instead? 

Also, the list is not parallel and I don't understand the technical details enough to rewrite it. Would you please check these guidelines https://developers.google.com/tech-writing/one/lists-and-tables

Author: @catalinaadam We can put bullet points.

2. Non-monotonic is nature.
3. It should be used in most of your queries.
4. Reads should be done from particular shards.
5. Writes should get written across all shards.

> Starting in MongoDB 5.0, you can reshard a collection by changing a collection's shard key. Follow the [MongoDB Documentation](https://www.mongodb.com/docs/manual/core/sharding-reshard-a-collection/#std-label-sharding-resharding) on how to change the shard key.

Reviewer 2: ```suggestion
Starting with MongoDB 5.0, you can reshard collections by changing their shard keys. For information on changing the shard key, check out [MongoDB documentation](https://www.mongodb.com/docs/manual/core/sharding-reshard-a-collection/#std-label-sharding-resharding).```

- **Clear the jumbo chunks:**

> To prevent the situation described above, we should check for jumbo chunks and remove the jumbo flag. Kindly refer to the [blog](https://www.percona.com/blog/finding-undetected-jumbo-chunks-in-mongodb/) on how to detect the jumbo chunks and split them.

Reviewer 2: ```suggestion
To prevent the situation described above, check for jumbo chunks and remove the jumbo flag. For information on detecting and splitting the jumbo chunks, see [Finding Undetected Jumbo Chunks in MongoDB
](https://www.percona.com/blog/finding-undetected-jumbo-chunks-in-mongodb/).
```

Reviewer 2: I would also replace "the situation described above" with the specific issue summary
